programs/vkd3d-hlsl-build/dxil_library.cpp

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2024 Philip Rebohle for Valve Software
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+#define INITGUID
+
+#include <array>
+#include <cstring>
+#include <iostream>
+
+#include "dxil_library.h"
+
+dxc_library dxc_library::instance;
+
+
+dxc_library::dxc_library()
+{
+    m_dxcompiler = LoadLibraryA("dxcompiler.dll");
+
+    if (!m_dxcompiler)
+    {
+        std::wcerr << "Failed to load dxcompiler.dll." << std::endl;
+        return;
+    }
+
+    m_dxc_create_instance = reinterpret_cast<DxcCreateInstanceProc>(
+        GetProcAddress(m_dxcompiler, "DxcCreateInstance"));
+
+    if (!m_dxc_create_instance)
+    {
+        std::wcerr << "DxcCreateInstance not found in dxcompiler.dll." << std::endl;
+        FreeLibrary(m_dxcompiler);
+        m_dxcompiler = nullptr;
+    }
+}
+
+
+dxc_library::~dxc_library()
+{
+    FreeLibrary(m_dxcompiler);
+}
+
+
+bool dxc_library::compile(const std::vector<char> &code, LPCWSTR source_name,
+        LPCWSTR target, LPCWSTR entry_point, std::vector<char> &binary) const
+{
+    if (!m_dxcompiler)
+        return false;
+
+    /* These are not thread-safe, just create new instances each time */
+    IDxcUtils *utils = nullptr;
+    HRESULT hr = m_dxc_create_instance(CLSID_DxcUtils, IID_PPV_ARGS(&utils));
+
+    if (FAILED(hr))
+    {
+        std::wcerr << "Failed to create IDxcUtils instance." << std::endl;
+        return false;
+    }
+
+    IDxcCompiler3 *compiler = nullptr;
+    hr = m_dxc_create_instance(CLSID_DxcCompiler, IID_PPV_ARGS(&compiler));
+
+    if (FAILED(hr))
+    {
+        std::wcerr << "Failed to create IDxcCompiler3 instance." << std::endl;
+        utils->Release();
+        return false;
+    }
+
+    IDxcIncludeHandler *include_handler = nullptr;
+    hr = utils->CreateDefaultIncludeHandler(&include_handler);
+
+    if (FAILED(hr))
+    {
+        std::wcerr << "Failed to create include handler." << std::endl;
+        utils->Release();
+        compiler->Release();
+        return false;
+    }
+
+    /* Set up parameters that minimize the generated binary size */
+    IDxcCompilerArgs *args = nullptr;
+
+    std::array<LPCWSTR, 2> arguments = {
+        L"-Qstrip_debug",
```

Comment:
u"" in case we port this to the Linux .so version of this at some unknown point?

Reply:
This produces compiler errors as casting char16_t* to wchar_t* implicitly does not work, this would need more ugly casts.

I'm not sure how useful it is to think about porting right now given that linux dxc does not produce signed binaries and we do not have a good enough alternative to d3dcompiler right now; if this ever becomes a concern we might as well revisit it then.

Reply:
Fair enough.